### ISSUE TITLE
Any tricks to lazily load dataset when creating Iterator?

### ISSUE THREAD

**Issue by JianyuZhan** [pytorch/text#176]
Currently OpenNMT-py encounters a memory hogging problem when using large corpus(size of several MBs), when totally loaded in to create `torchtext.data.Dataset`, it will occupy ~GBs memory.

Our (proposed) approach is to divide the large corpus into small ones, and create `torchtext.data.Dataset` for each separately when preprocessing. This approach works fine.

But when do training, this problem emerges again. Because we need to create `torchtext.data.Iterator`, which takes `dataset` as argument. Two possible approaches:

1)  Coalesce multiple datasets into one and pass to `torchtext.data.Iterator`. This approach requires all
    datasets loaded at the same time, so the memory hogging problem comes back.
2)  Create  each `torchtext.data.Iterator` for each `dataset`, and make an iterator chain to iterate over all data.  This actually also requires dataset to be loaded when the Iterator is created, unless we postpone the creation of `Iterator`. And it is hard to iterate across the dataset boundary, because our current `__iter__` yields a wrapper `Batch`.

Any tricks to tackle this problem?  I figure we could add support for accepting multiple datasets and do lazy load when create Iterator.  Thought?  Thanks.

**Comment by jekbradbury** [comment]
The original trick for a lazy iterator was to make the `Dataset` a Python generator (implemented however you want) and make sure to use an `Iterator` without (global) shuffling or sorting (so for instance `BucketIterator` with `sort=False` and `shuffle=False`). Then what will happen is that the `BucketIterator` or user equivalent will prefetch some number of examples from the lazy `Dataset` and perform the bucketing and batching on the prefetched examples. That might be a third option here?

**Comment by jekbradbury** [comment]
Here's a demo of how that works: https://drive.google.com/file/d/11yqYCQJiwIevZguXl7x9Oo6dpl7keRpy/view?usp=sharing

**Comment by Rhuax** [comment]
Hi, I would like to reopen this issue since I'm facing the same problem. I have a training set about 4.5 GB and using the standard way of creating a BucketIterator fills up all of my ram capacity (12GB). 
Following the example posted by @jekbradbury he is generating new samples on the fly by randomly picking words, but how can I actually feed real samples from a huge file?

**Comment by Rhuax** [comment]
Hi, I would like to reopen this issue since I'm facing the same problem. I have a training set about 4.5 GB and using the standard way of creating a BucketIterator fills up all of my ram capacity (12GB). 
Following the example posted by @jekbradbury he is generating new samples on the fly by randomly picking words, but how can I actually feed real samples from a huge file?

**Comment by Rhuax** [comment]
Using @jekbradbury 's demo, I cannot use the iterator more than once. Is there something I can do? I'm using a bucketIterator to iterate over test set examples, so I need to iterate over it more than once during the training phase, but after iterating once, from the second time on the iterator doesn't let me iterate. What can I do?
